File: src/commands/ping.ts
import type { JoplinClient } from '../client.js';
import type { Command } from '../lib/command.js';
import type { CommandSpec, ParsedArgs } from '../lib/args.js';
import { object } from '../toon.js';

const spec: CommandSpec = {
  name: 'ping',
  summary: 'Check connectivity and authentication against the Joplin Web Clipper API.',
  usage: 'joplin-axi ping',
  flags: {},
  examples: ['joplin-axi ping'],
};

async function run(_parsed: ParsedArgs, client: JoplinClient): Promise<string> {
  const clipper = await client.ping();
  let auth = 'failed';

  if (clipper) {
    try {
      await client.listNotebooks();
      auth = 'ok';
    } catch {
      auth = 'failed';
    }
  }

  return object('ping', { clipper: clipper ? 'reachable' : 'unreachable', auth });
}

export const pingCommand: Command = { spec, run };

File: internal/commands/ping.go
package commands

import (
	"context"

	"github.com/gcerar/joplin-axi/internal/args"
	"github.com/gcerar/joplin-axi/internal/client"
	"github.com/gcerar/joplin-axi/internal/toon"
)

var pingSpec = args.CommandSpec{
	Name:     "ping",
	Summary:  "Check connectivity and authentication against the Joplin Web Clipper API.",
	Usage:    "joplin-axi ping",
	Flags:    nil,
	Examples: []string{"joplin-axi ping"},
}

func runPing(ctx context.Context, _ args.ParsedArgs, c client.Client) (CommandResult, error) {
	clipperOK := c.Ping(ctx)
	auth := "failed"

	if clipperOK {
		if _, err := c.ListNotebooks(ctx, nil); err == nil {
			auth = "ok"
		}
	}

	clipperStatus := "unreachable"
	if clipperOK {
		clipperStatus = "reachable"
	}

	return Ok(toon.Object("ping", []toon.Field{
		{Key: "clipper", Value: clipperStatus},
		{Key: "auth", Value: auth},
	})), nil
}

// PingCommand checks connectivity and authentication against the Joplin
// Web Clipper API.
var PingCommand = Command{Spec: pingSpec, Run: runPing}
